src/lib/keccak.ts

```diff
@@ -209,18 +389,233 @@ function permutation(state: Field[][], rc: Field[]): Field[][] {
+
+  for (let i = 0; i < paddedMessage.length; i += rate / 8) {
+    // split into blocks of rate bits
+    // for each block of rate bits in the padded message -> this is rate/8 bytes
+    const block = paddedMessage.slice(i, i + rate / 8);
+    // pad the block with 0s to up to 1600 bits
+    const paddedBlock = block.concat(zeros);
+    // padded with zeros each block until they are 1600 bit long
+    assert(
+      paddedBlock.length * 8 === KECCAK_STATE_LENGTH,
+      'improper Keccak block length'
+    );
+    const blockState = getKeccakStateOfBytes(paddedBlock);
+    // xor the state with the padded block
+    const stateXor = keccakStateXor(state, blockState);
+    // apply the permutation function to the xored state
+    const statePerm = permutation(stateXor, rc);
+    state = statePerm;
+  }
+  return state;
+}
+
+// Squeeze state until it has a desired length in bits
+function squeeze(
+  state: Field[][],
+  length: number,
+  rate: number,
+  rc: Field[]
+): Field[] {
+  // Copies a section of bytes in the bytestring into the output array
+  const copy = (
+    bytestring: Field[],
+    outputArray: Field[],
+    start: number,
+    length: number
+  ) => {
+    for (let i = 0; i < length; i++) {
+      outputArray[start + i] = bytestring[i];
+    }
+  };
+
+  let newState = state;
+
+  // bytes per squeeze
+  const bytesPerSqueeze = rate / 8;
+  // number of squeezes
+  const squeezes = Math.floor(length / rate) + 1;
+  // multiple of rate that is larger than output_length, in bytes
+  const outputLength = squeezes * bytesPerSqueeze;
+  // array with sufficient space to store the output
+  const outputArray = Array(outputLength).fill(Field.from(0));
+  // first state to be squeezed
+  const bytestring = keccakStateToBytes(state);
+  const outputBytes = bytestring.slice(0, bytesPerSqueeze);
+  copy(outputBytes, outputArray, 0, bytesPerSqueeze);
+
+  // for the rest of squeezes
+  for (let i = 1; i < squeezes; i++) {
+    // apply the permutation function to the state
+    newState = permutation(newState, rc);
+    // append the output of the permutation function to the output
+    const bytestringI = keccakStateToBytes(state);
+    const outputBytesI = bytestringI.slice(0, bytesPerSqueeze);
+    copy(outputBytesI, outputArray, bytesPerSqueeze * i, bytesPerSqueeze);
+  }
+
+  // Obtain the hash selecting the first bitlength/8 bytes of the output array
+  const hashed = outputArray.slice(0, length / 8);
+  return hashed;
+}
+
+// Keccak sponge function for 1600 bits of state width
+// Need to split the message into blocks of 1088 bits.
+function sponge(
+  paddedMessage: Field[],
+  length: number,
+  capacity: number,
+  rate: number
+): Field[] {
+  // check that the padded message is a multiple of rate
+  if ((paddedMessage.length * 8) % rate !== 0) {
+    throw new Error('Invalid padded message length');
+  }
+
+  // load round constants into Fields
+  let rc = exists(24, () => TupleN.fromArray(24, ROUND_CONSTANTS));
```

Comment:
I know this is how it was done in OCaml, but I wonder if it would make sense to defer the creation of `rc` until `permutation` function. It looks a bit unnecessary to be passing it to both absorb and squeeze at the first level.